chartfx-dataset/src/main/java/de/gsi/dataset/serializer/IoClassSerialiser.java

```diff
@@ -0,0 +1,458 @@
+        final String subClass = classField.getFieldName();
+        if (recursionDepth != 0 && startMarkerFunction != null) {
+            startMarkerFunction.accept(subClass);
+        }
+
+        final Object newRoot = classField.getField() == null ? rootObj : classField.getField().get(rootObj);
+        for (final FieldDescription fieldDescription : classField.getChildren()) {
+            ClassFieldDescription field = (ClassFieldDescription) fieldDescription;
+
+            if (!field.isPrimitive()) {
+                final Object reference = field.getField().get(newRoot);
+                if (!field.isPrimitive() && reference == null) {
+                    // only follow and serialise non-null references of sub-classes
+                    continue;
+                }
+            }
+            serialiseObject(newRoot, field, recursionDepth + 1);
+        }
+
+        if (recursionDepth != 0 && endMarkerFunction != null) {
+            endMarkerFunction.accept(subClass);
+        }
+    }
+
+    public void serialiseObject(final Object obj) {
+        if (obj == null) {
+            // serialise null object
+            ioSerialiser.putHeaderInfo();
+            ioSerialiser.putEndMarker("OBJ_ROOT_END");
+            return;
+        }
+
+        final ClassFieldDescription classField = ClassUtils.getFieldDescription(obj.getClass());
+        final FieldSerialiser<?> existingSerialiser = classField.getFieldSerialiser();
+        final FieldSerialiser fieldSerialiser = existingSerialiser == null ? cacheFindFieldSerialiser(classField.getType(), classField.getActualTypeArguments()) : existingSerialiser;
+
+        if (fieldSerialiser == null) {
+            ioSerialiser.putHeaderInfo(classField);
+            serialiseObject(obj, classField, 0);
+            ioSerialiser.putEndMarker(classField.getFieldName());
+        } else {
+            if (classField.getFieldSerialiser() == null) {
+                classField.setFieldSerialiser(fieldSerialiser);
+            }
+            ioSerialiser.putHeaderInfo();
+            ioSerialiser.putCustomData(classField, obj, obj.getClass(), fieldSerialiser);
+            ioSerialiser.putEndMarker("OBJ_ROOT_END");
+        }
+    }
+
+    protected boolean checkClassCompatibility(final List<Type> ref1, final List<Type> ref2) {
+        if (ref1.size() != ref2.size()) {
+            return false;
+        }
+        if (ref1.isEmpty() && ref2.isEmpty()) {
+            return true;
+        }
+
+        for (int i = 0; i < ref1.size(); i++) {
+            final Class<?> class1 = ClassUtils.getRawType(ref1.get(i));
+            final Class<?> class2 = ClassUtils.getRawType(ref2.get(i));
+            if (!class1.equals(class2) && !(class2.isAssignableFrom(class1))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    protected void deserialise(final Object obj, final FieldDescription fieldRoot, final ClassFieldDescription classField, final int recursionDepth) {
```

Comment:
Codacy found an issue: [The method 'deserialise(Object, FieldDescription, ClassFieldDescription, int)' has an NPath complexity of 1092, current threshold is 200](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5905529)